--- /etc/sysconfig/init	2013-11-21 09:51:28.000000000 -0500
+++ sources/genesis-scripts/sysconfig-init	2014-04-10 21:41:10.000000000 -0400
@@ -20,7 +20,7 @@
 # Set to 'yes' to allow probing for devices with swap signatures
 AUTOSWAP=no
 # What ttys should gettys be started on?
-ACTIVE_CONSOLES=/dev/tty[1-6]
+ACTIVE_CONSOLES=/dev/tty1
 # Set to '/sbin/sulogin' to prompt for password on single-user mode
 # Set to '/sbin/sushell' otherwise
 SINGLE=/sbin/sushell
